compiler/damlc/daml-doc/src/DA/Daml/Doc/Render.hs

```diff
@@ -5,49 +5,112 @@
 import qualified Data.Aeson.Encode.Pretty as AP
 import qualified Data.Text as T
-import qualified Data.Text.Lazy as TL
-import qualified Text.Blaze.Html5 as H
-import qualified Text.Blaze.Html.Renderer.Text as H
+import qualified Data.Text.Encoding as T
+import qualified Data.ByteString as BS
+import qualified Data.Map.Strict as Map
 
 
 -- | centralised JSON configuration for pretty-printing
 jsonConf :: AP.Config
 jsonConf = AP.Config (AP.Spaces 2) (AP.keyOrder ["id"]) AP.Generic True
 
+renderDocs :: RenderOptions -> [ModuleDoc] -> IO ()
+renderDocs RenderOptions{..} mods = do
+    let renderModule =
+            case ro_format of
+                Json -> const (renderLine "") -- not implemented (yet?)
+                Hoogle -> const (renderLine "") -- not implemented (yet?)
+                Rst -> renderSimpleRst
+                Markdown -> renderSimpleMD
+                Html -> renderSimpleMD
+        postProcessing =
+            case ro_format of
+                Html -> GFM.commonmarkToHtml [GFM.optUnsafe] [GFM.extTable]
+                _ -> id
+        template = fromMaybe (defaultTemplate ro_format) ro_template
+
+    case ro_mode of
+        RenderToFile path -> do
+            BS.writeFile path
+                . T.encodeUtf8
+                . renderTemplate template
+                    (fromMaybe "Package Docs" ro_title)
+                . postProcessing
+                . renderPage
+                $ mconcatMap renderModule mods
+
+        RenderToFolder path -> do
+            let renderMap = Map.fromList
+                    [(md_name mod, renderModule mod) | mod <- mods]
+                outputMap = renderFolder renderMap
+                extension =
+                    case ro_format of
+                        Json -> "json"
+                        Hoogle -> "txt"
+                        Markdown -> "md"
+                        Rst -> "rst"
+                        Html -> "html"
+
+                outputPath mod = path </> moduleNameToFileName mod <.> extension
+                pageTitle mod = T.concat
+                    [ maybe "" (<> " - ") ro_title
+                    , "Module "
+                    , unModulename mod ]
 
--- TODO rendering structures closely resembles each other. Could share code and
--- use a common typeclass if need be (unsure about ROI so far).
+            createDirectoryIfMissing True path
+            for_ (Map.toList outputMap) $ \ (mod, renderedOutput) -> do
+                BS.writeFile (outputPath mod)
+                    . T.encodeUtf8
+                    . renderTemplate template (pageTitle mod)
+                    . postProcessing
+                    $ renderedOutput
 
-data DocFormat = Json | Rst | Markdown | Html | Hoogle
-  deriving (Eq, Show, Read, Enum, Bounded)
 
+renderTemplate ::
+    T.Text -- ^ template
+    -> T.Text -- ^ page title
+    -> T.Text -- ^ page body
+    -> T.Text
+renderTemplate template pageTitle pageBody
+    = T.replace "__BODY__" pageBody
```

Comment:
If this starts becoming more complex we might want to switch to some standard templating library, e.g., `mustache`. For now this is totally fine.

Reply:
Agreed :-)